dell/csm-replication
# Integrate PVC remap feature to replication controller for single cluster failover

# Description
- The feature will detect a successful failover on a single cluster, and remap the direction of PV pairs for all PVCs under the same replication group. 
- Expected result after a planned/unplanned failover: The old PVC will be deleted and a new PVC with the corresponding remote PV, remote storage class name, and the remote replication group. The reclaim policy of both PVs will stay the same.
- The feature can be disabled by adding an argument flag `--disable-pvc-remap=True` to the replication controller deployment config `helm-charts/charts/csm-replication/templates/controller.yaml` when deploying with helm chart, as shown below:
![image](https://github.com/dell/csm-replication/assets/171162480/dbcd2cf6-afd4-4a11-995b-c8104cf05e3c)
Alternatively, you can edit the replication controller after its deployment, and add that argument inside the yaml file with `k edit -n dell-replication-controller deployment`.

# GitHub Issues
List the GitHub issues impacted by this PR:
| GitHub Issue # |
| -------------- |
|N/A |

# Checklist:


# How Has This Been Tested?
- [x] Manual test planned failover for a replication group with 40 PVCs, costing 2m. Manual test unplanned failover.
- [x] Unit testing for planned, unplanned, disable pvcremap, pvc remote pv mismatched.
- [x] Integrated testing for init, synchronize, failover with pvc remap, reprotect, and cleanup.

# Manual Testing Instruction
- Deploy an app. Check if PV, PVC, and RG show up as expected. For example,
![image](https://github.com/dell/csm-replication/assets/171162480/2005df3d-7d56-46c4-add9-3e6bf1972507)
- Helm delete the app deployment pod.
- Perform a failover: `repctl --rg <rg-id> failover --target <replicated-rg-id>`
- The expected result: The Reclaim Policy for each PV should be what it originally was before failover and the PVC should be bound to the replicated PV. 
![vd2gzgag](https://github.com/dell/csm-replication/assets/171163262/034b753f-e99d-46ec-9121-a21266b4ad20)
- Note that the IsSource is not reflected until reprotect. To perform a reprotect: `repctl --rg <replicated-rg-id> reprotect`
- To go back to the original state, run a second failback (with arguments switcher) and reprotect (on primary rg).


# Test Coverage Change
![image](https://github.com/user-attachments/assets/f68e33ae-087a-4f82-b66a-8072106ce397)

`/controllers/replication-controller/dellcsireplicationgroup_controller.go` change:
| Original test coverage of main branch |  Latest test coverage due to added code |
| -------------- |  -------------- |
|58.4%             | 68.2%             |
